20170827 switch dynDNS handling to RRset model

peterthomassen opened the pull request:
This PR also introduces the RR model for record contents and, as a consequence, comes with a rewrite of the RRset serializers. It finally allows us to get rid of the nasty serializer local imports in various locations.

It also fixes a few other minor things (see commit list).

peterthomassen commented:
I still have to go through the last commit once more to see if things follow KISS. The other commits are ready for review though.

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -219,6 +218,9 @@ def save(self, *args, **kwargs):
 
         self.pdns_sync(new_domain)
 
+    def __str__(self):
```
My guess is this is needed for the `StringRelatedField` on the `RRSet` serializer, so we should add a note here that this value must be unique

peterthomassen replied:
done

nils-wisiol commented on api/desecapi/serializers.py:
```diff
@@ -1,45 +1,53 @@
+        return super().update(instance, validated_data)
+
+    def get_records(self, obj):
+        return [x for x in obj.records.values_list('content', flat=True)]
```
order is probably not guaranteed here, should we sort?

nils-wisiol replied:
Also, `[x for x in ...]` is an unnecessary list comprehension. Why not use the list directly? If it is only an iterable, it can easily converted into a list using `list()`

peterthomassen replied:
done, using `list()`

peterthomassen replied:
not sorting for now, as the ideal order would be 1) unclear and 2) type-dependent

nils-wisiol commented on api/desecapi/serializers.py:
```diff
@@ -1,45 +1,53 @@
+    def _inject_records_data(self, validated_data):
+        records_data = [{'content': x}
+                        for x in self.context['request'].data['records']]
+        rrs = RRSerializer(data=records_data, many=True, allow_empty=False)
```
will this cause `CREATE` statements to the database?

peterthomassen replied:
You probably mean `INSERT`. No, we are just feeding user input to the serializer to use the validation capabilites.

nils-wisiol commented on api/desecapi/serializers.py:
```diff
@@ -1,45 +1,53 @@
+        return [x for x in obj.records.values_list('content', flat=True)]
+
+    def validate_type(self, value):
+        if value in RRset.RESTRICTED_TYPES:
```
does this cover lowercase types as well?

peterthomassen replied:
`type` is required to be uppercase by the validator attached to the `RRset.type` field definition (it's checked by the `ModelSerializer`)

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -156,56 +134,91 @@ def pdns_id(self):
+                                      type=rrset.type)
+
+            # Identical RRset exists; no need to recreate
+            if RR.objects.filter(rrset__in=qs, rrset__ttl=rrset.ttl).values('content') == rrset.records_data:
```
may have false positives due to ordering

peterthomassen replied:
fixed

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -156,56 +134,91 @@ def pdns_id(self):
+            qs = RRset.objects.filter(domain=self.pk, subname=rrset.subname,
+                                      type=rrset.type)
+
+            # Identical RRset exists; no need to recreate
```
"if identical RRset ...."

peterthomassen replied:
done

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -156,56 +134,91 @@ def pdns_id(self):
+        # RRset.objects.bulk_create(rrsets)
+
+    @transaction.atomic
+    def set_rrsets(self, rrsets):
```
add method comment

peterthomassen replied:
done

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -156,56 +134,91 @@ def pdns_id(self):
+            # Only clean up direct children of dedyn.io
+            if parent_pdns_id == 'dedyn.io.':
+                # Can't just call delete() on the queryset (would miss RRset.delete())
+                rrsets = RRset.objects.filter(domain__name='dedyn.io',
```
`domain_name`? (one _)

peterthomassen replied:
nope, this is a related field selector

nils-wisiol commented on api/desecapi/views.py:
```diff
@@ -362,9 +362,16 @@ def get(self, request, format=None):
+        rrsets = []
+        datas = {'A': self.findIPv4(request), 'AAAA': self.findIPv6(request)}
+
+        for type, ip in datas.items():
```
hah! this shadows type xD

peterthomassen replied:
done

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -281,6 +282,7 @@ class Meta:
         unique_together = (("domain","subname","type"),)
 
     def __init__(self, *args, **kwargs):
+        self.records_data = kwargs.pop('records_data', [])
```
done (switching `*args` and `records_data`, though)

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -337,18 +333,31 @@ def update_pdns(self):
+
+class RR(models.Model):
+    rrset = models.ForeignKey(RRset, on_delete=models.CASCADE, related_name='records')
+    # https://lists.isc.org/pipermail/bind-users/2008-April/070148.html
```
done

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -337,18 +333,31 @@ def update_pdns(self):
 
-        if self._dirty or new:
+        records = self.records.all()
+        if self.records_data and self.records_data != [{'content': x.content}
```
done

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -337,18 +333,31 @@ def update_pdns(self):
+            self._dirty = True
+            records.delete()
+            while self.records_data:
+                self.records.create(**self.records_data.pop())
```
done

peterthomassen commented on api/desecapi/models.py:
```diff
@@ -192,14 +192,13 @@ def pdns_sync(self, new_domain):
+            if rrset['type'] not in RRset.RESTRICTED_TYPES:
+                RRset(**rrset).save(pdns=False)
 
-    @transaction.atomic
```
dealt with

peterthomassen commented:
I addressed all your review comments and improved a few other things I had on my mind with commit https://github.com/desec-io/desec-stack/pull/69/commits/b97c1d4c2f23618cf53bd7be2b395ca10aa7c882. I don't have any other todo items, so please take a look at it again.

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -143,35 +143,35 @@ def sync_to_pdns(self):
+                       if rrset_data['type'] not in RRset.RESTRICTED_TYPES]
+        # Can't do bulk create because we need records creation in RRset.save()
+        for rrset_data in rrset_datas:
+            RRset(**rrset_data).save(sync=False)
```
potential run time optimization:

- first create all `RRset` objects
- bulk insert all RRset objects
- collect all dependent `RR` objects
- bulk insert all `RR` objects

that should reduce overhead to two DB transactions

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -181,44 +181,48 @@ def set_rrsets(self, rrsets):
+            except RRset.DoesNotExist:
+                pass
+
+            # At this point, rrset is an RRset to be created or to be updated.
```
or neither?

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -181,44 +181,48 @@ def set_rrsets(self, rrsets):
+                    rrset.records_data = []
+
+                parent.set_rrsets(rrsets)
+                rrsets.delete()
```
unnecessary? `set_rrsets` calls delete for empty record sets

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -324,54 +330,71 @@ def setter_type(self, val):
+    def save(self, sync=True, *args, **kwargs):
         new = self.pk is None
         self.updated = timezone.now()
         self.full_clean()
```
move above two lines into `if new or 'ttl' in self.get_dirties():` below

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -386,7 +386,8 @@ def save(self, sync=True, *args, **kwargs):
 
-        if sync and (new or self.get_dirties()):
+        if sync and not self.domain.owner.captcha_required \
+                and (new or 'records' in self.get_dirties()):
```
`'ttl' in self.get_dirties()` constitutes also an update condition

nils-wisiol commented on api/desecapi/management/commands/sync-from-pdns.py:
```diff
@@ -20,12 +20,13 @@ def handle(self, *args, **options):
                     raise CommandError('{} is not a known domain'.format(domain_name))
 
         for domain in domains:
+            print(domain.name, '...', end='', flush=True)
```
use `self.stdout.write` like in the [django docs example ](https://docs.djangoproject.com/en/1.11/howto/custom-management-commands/)

nils-wisiol commented on api/desecapi/authentication.py:
```diff
@@ -98,3 +93,20 @@ def authenticate_credentials(self, userid, key):
+    Authentication against remote IP address for dedyn.io management by nslord
+    """
+    def authenticate(self, request):
+        nslord = '::ffff:%s.1.11' % os.environ['DESECSTACK_IPV4_REAR_PREFIX16']
```
is this dual-stack safe?

nils-wisiol commented on api/desecapi/views.py:
```diff
@@ -15,8 +15,7 @@
 from django.template.loader import get_template
-from desecapi.authentication import (
-    BasicTokenAuthentication, URLParamAuthentication)
+from desecapi.authentication import *
```
[explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/) xD

nils-wisiol commented on api/desecapi/views.py:
```diff
@@ -164,6 +163,7 @@ def update(self, request, *args, **kwargs):
 
 
 class RRsetList(generics.ListCreateAPIView):
+    authentication_classes = (TokenAuthentication, IPAuthentication,)
```
extent default list rather than overriding it

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -103,6 +103,8 @@ class Domain(models.Model, mixins.SetterMixin):
     owner = models.ForeignKey(settings.AUTH_USER_MODEL, related_name='domains')
     _dirtyName = False
+    _ns_records_data = [{'content': 'ns1.desec.io.'},
+                        {'content': 'ns2.desec.io.'}]
```
move to settings.py, possibly read from .env

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -233,8 +250,19 @@ def save(self, *args, **kwargs):
+            # If the domain is a direct subdomain of dedyn.io, set NS records
+            # in parent. Don't notify slaves (we first have to enable DNSSEC).
+            if parent_pdns_id == 'dedyn.io.':
+                parent = Domain.objects.filter(name='dedyn.io').first()
```
split into two atomic parts

nils-wisiol commented on api/desecapi/models.py:
```diff
@@ -365,7 +393,7 @@ def delete(self, *args, **kwargs):
 
     @transaction.atomic
-    def save(self, sync=True, *args, **kwargs):
+    def save(self, sync=True, notify=True, *args, **kwargs):
```
do not forward `sync`, `notify` to `super()`

nils-wisiol commented on api/desecapi/views.py:
```diff
@@ -55,10 +55,15 @@ def perform_create(self, serializer):
 
-        queryset = Domain.objects.filter(Q(name=domain_list[0]) | (Q(name__in=domain_list[1:]) & ~Q(owner=self.request.user)))
+        # Remove public suffixes and then use this list to control registration
+        public_suffixes = {'dedyn.io'}
```
make this safe for two-dot public suffixes (like co.uk)

nils-wisiol commented on nslord/cronhook/secure-zones.sh:
```diff
@@ -13,19 +13,21 @@ for ZONE in `echo "SELECT name FROM domains WHERE type = 'NATIVE' && id NOT IN(S
-		DATA='{"rrsets": [ {"name": "'"$ZONE".'", "type": "DS", "ttl": 60, "changetype": "REPLACE", "records": '
+		echo "Getting DS records for $ZONE and put them in parent zone"
+		DATA='{"subname": "'"$SUBNAME"'", "type": "DS", "ttl": 60, "records": '
 		DATA+=`curl -sS -X GET -H "X-API-Key: $APITOKEN" http://nslord:8081/api/v1/servers/localhost/zones/$ZONE/cryptokeys \
```
quote URL
